### ISSUE TITLE
two questions on parallel implementation in the batch-A3C

### ISSUE THREAD

**Issue by Emerald01** [gliese581gg/batch-A3C_tensorflow#3]
Hello,

I looked at your bA3C code and learned a lot. Thank you so much for sharing your codebase implementing this idea. I have two questions mainly on your code. BTW, I do not think your code has any problem, I just have a few speed concerns that want to discuss with you a little bit.

(1) python multi-threading is known to be slow for CPU-bound problem because of GIL. I am not sure whether this embedded python thread locker would slow down the job although GPU is the main executer within the worker, at least I guess gym environment in charge of playing the game seems to constantly use CPU.

(2) for parallel env exploration, since the policy and value network is globally shared, would GPU performance be reduced due to the concurrent visits by different workers? 

Best,
Tian

**Comment by gliese581gg** [comment]
Hi!

(1)
You are right. I had no idea about GIL when I implement this.

Now I am using multiprocessing and Pytorch, which supports shared cuda memory across processes.

I will upload pytorch version when I have a time:)

(2)
I have not thought about this. Maybe I have to test whether shared network slows down learning in multi processing

**Comment by Emerald01** [comment]
Cool, I am glad that we have similar concerns :)

BTW, as for A3C algorithm itself, I have a doubt. I actually want to ask the author of the original A3C paper but did not get a reply.

Say we have two envs running in parallel, env1 and env2, they use a common neural network model denoted by Global_NN to explore and write the buffer with its own experience.  With those experiences, Global_NN updates. 

Writing in a pseudo code, I think the training loop is roughly like this:

# say env1 gets "exp1", env2 gets "exp2"
exp1 = explore(env1, Global_NN), writes to buffer when finished
exp2 = explore(env2, Global_NN), writes to buffer when finished

#2 Global_NN fetches every exp from buffer (I just assume here the order is [exp1, exp2])  
Global_NN_new = train(exp1, Global_NN)
Global_NN_new2 = train(exp2, Global_NN_new)

My concern is how to understand the validity of this async training procedure: Global_NN becomes Global_NN_new because of the training from exp1, therefore when exp2 is fetched to update model, it is actually applied to  Global_NN_new instead of Global_NN. However, exp2 is based on the original Global_NN. We basically apply a gradient explored from an apple to a pear. I feel the most rigorous approach is that we start to explore and obtain exp2 after the model is updated by exp1, but that is just a sequential exploration. 

I believe the A3C is very sensitive to how large every training exp would be, if the training size is small so every change on the Global_NN is small, it could stabilize the optimization. Otherwise I think we should get grad_1 from exp1 and grad_2 from exp2, then average them out to apply to the Global_NN. Essentially I feel it would make more sense to make an async exploration but a synchronized learning by averaging.

**Comment by Emerald01** [comment]
Cool, I am glad that we have similar concerns :)

BTW, as for A3C algorithm itself, I have a doubt. I actually want to ask the author of the original A3C paper but did not get a reply.

Say we have two envs running in parallel, env1 and env2, they use a common neural network model denoted by Global_NN to explore and write the buffer with its own experience. With those experiences, Global_NN updates.

Writing in a pseudo code, I think the training loop is roughly like this:

// say env1 gets "exp1", env2 gets "exp2"

exp1 = explore(env1, Global_NN), writes to buffer when finished
exp2 = explore(env2, Global_NN), writes to buffer when finished

// Global_NN fetches every exp from buffer (I just assume here the order is [exp1, exp2])
Global_NN_new = train(exp1, Global_NN)
Global_NN_new2 = train(exp2, Global_NN_new)

My concern is how to understand the validity of this async training procedure: Global_NN becomes Global_NN_new because of the training from exp1, therefore when exp2 is fetched to update model, it is actually applied to Global_NN_new instead of Global_NN. However, exp2 is based on the original Global_NN. We basically apply a gradient explored from an apple to a pear. I feel the most rigorous approach is that we start to explore and obtain exp2 after the model is updated by exp1, but that is just a sequential exploration.

I believe the A3C is very sensitive to how large every training exp would be, if the training size is small so every change on the Global_NN is small, it could stabilize the optimization. Otherwise I think we should get grad_1 from exp1 and grad_2 from exp2, then average them out to apply to the Global_NN. Essentially I feel it would make more sense to make an async exploration but a synchronized learning by averaging.

**Comment by Emerald01** [comment]
Cool, I am glad that we have similar concerns :)
BTW, I tried Cython, it has the way to release the GIL so it can provide a real multi-threading env, but all codes need to rewrite in the C grammar.
